### PR TITLE
Pass a project when creating an invisible editor

### PR DESCRIPTION
**Pull request by sharwell**
This change substantially reduces the amount of work required to locate a document within the solution (file path to hierarchy+itemid).

* Time to apply Fix All from #19895: 2:45 mins
* Time to apply Fix All after this change: 1:32 mins

## Ask Mode

**Customer scenario**

A customer applies a Fix All operation which affects many files in a large solution. (Not specific to any particular code fix.)

**Bugs this fixes:**

Derived from #19900 as the safe option for 15.3 (provides some but not all possible benefits).

**Workarounds, if any**

Wait longer.

**Risk**

Low. This feature adds a hint for a lookup, but valid information is not required in order for the underlying operation to succeed (falls back to slow case when information is not available).

**Performance impact**

This is a substantial performance improvement for Fix All operations.

**Is this a regression from a previous update?**

No.

**Root cause analysis:**

No performance tests covering large-scale operations (yet).

**How was the bug found?**

Internal testing.

### CODE REVIEW

**Comment by sharwell** [comment]
@Pilchie Yes it works fine. Tested on vs-threading which targets 3 frameworks:

```xml
<TargetFrameworks>netstandard1.1;portable-net45+win8+wpa81;net45</TargetFrameworks>
```

File: src/VisualStudio/Core/Def/Implementation/ProjectSystem/InvisibleEditor.cs [hunk 1]
```diff
@@ -28,15 +28,16 @@ internal partial class InvisibleEditor : IInvisibleEditor
...
         {
             _serviceProvider = serviceProvider;
             _filePath = filePath;
             _needsSave = needsSave;
 
             var invisibleEditorManager = (IIntPtrReturningVsInvisibleEditorManager)serviceProvider.GetService(typeof(SVsInvisibleEditorManager));
+            var vsProject = projectOpt?.Hierarchy as IVsProject;
             var invisibleEditorPtr = IntPtr.Zero;
-            Marshal.ThrowExceptionForHR(invisibleEditorManager.RegisterInvisibleEditor(filePath, null, 0, null, out invisibleEditorPtr));
+            Marshal.ThrowExceptionForHR(invisibleEditorManager.RegisterInvisibleEditor(filePath, vsProject, 0, null, out invisibleEditorPtr));
```
Comment (CyrusNajmabadi) [inline src/VisualStudio/Core/Def/Implementation/ProjectSystem/InvisibleEditor.cs]: So clearly this helps, but why?  Can we document this code to make it clear why this is necessary?

**Comment by davkean** [comment]
The hierarchy we're passing you implements IVsHierarchy and IVsProject4, can we confirm that RegisterInvisibleEditor isn't looking beyond those?

**Comment by jasonmalinowski** [comment]
@davkean: I'll let @sharwell confirm the answer, but I think this might actually mean your requirements are weaker. By passing null the shell goes and asks all projects for the stuff. It might just skip that and assume nothing further here.

Ironically this was one workaround I considered for dotnet/project-system#2160 but didn't and filed the over there not thinking there were other benefits to it. But the perf wins are obviously goodness!

**Comment by davkean** [comment]
@jasonmalinowski We always implement IVsProject so the cast will succeed. I'm more worried about what it's sniffing under the covers? I assuming it's doing a IsDocumentInProject call but would like confirmation .

**Comment by sharwell** [comment]
@davkean Two things:

1. `IVsProject4` extends `IVsProject` (it looks like that was implied above but wanted to make it explicit)
2. If you fail to pass a project, the code in `CEditorManager::GetDocumentHierarchyAndItemID` goes through `IVsUIShellOpenDocument4::IsDocumentInAProject2`, which looks at *everything*. If you pass a project, it goes through `IVsProject::IsDocumentInProject`.

**Comment by davkean** [comment]
Thanks, that addresses my concern.

**Comment by sharwell** [comment]
@davkean @CyrusNajmabadi @jasonmalinowski Updated the documentation to explain why the new code makes such a difference.

**Comment by sharwell** [comment]
@MattGertz for ask mode

**Comment by jasonmalinowski** [comment]
@sharwell And if we passed a project that the project system doesn't think is in it, what happens? Does it fail, or does it fall back and do the full loop as if you had passed null? The answer doesn't change my signoff, but just to understand if we see any surprises here.

**Comment by sharwell** [comment]
➡️ Filed #19952 to follow up on @jasonmalinowski's investigation request.

**Pull request merged**